packages/grid/src/vaadin-grid-sort-mixin.js

```diff
@@ -122,44 +122,27 @@ export const SortMixin = (superClass) =>
       }
     }
 
-    /** @private */
-    __appendSorter(sorter) {
-      if (!sorter.direction) {
-        this._removeArrayItem(this._sorters, sorter);
-      } else if (!this._sorters.includes(sorter)) {
-        this._sorters.push(sorter);
-      }
-    }
-
-    /** @private */
-    __prependSorter(sorter) {
-      this._removeArrayItem(this._sorters, sorter);
-      if (sorter.direction) {
-        this._sorters.unshift(sorter);
-      }
-    }
-
     /** @private */
     __updateSorter(sorter, shiftClick, fromSorterClick) {
-      if (!sorter.direction && this._sorters.indexOf(sorter) === -1) {
+      if (!sorter.direction && !this._sorters.includes(sorter)) {
         return;
       }
 
       sorter._order = null;
 
+      const restSorters = this._sorters.filter((s) => s !== sorter);
       if (
         (this.multiSort && (!this.multiSortOnShiftClick || !fromSorterClick)) ||
         (this.multiSortOnShiftClick && shiftClick)
       ) {
         if (this.multiSortPriority === 'append') {
-          this.__appendSorter(sorter);
+          this._sorters = [...restSorters, sorter];
```

Comment:
The previous implementation seems to avoid assigning a new value for the `_sorters` array. There's an observer that reacts to `_sorters` change in array-data-provider-mixin.

Reply:
Yeah, but the current implementation already does assign a new value in single column sorting:

https://github.com/vaadin/web-components/pull/6981/files#diff-3c17927c07b9009af8f21234efdca93c79dfb5095cb182cff0381a875074ebcaR144

So, I assumed that it shouldn't be a big deal to do so in multi column sorting too.

Reply:
By the way looks like the reason `ArrayDataProviderMixin` subscribes to `_sorters` and `_filters` is not to trigger a data refresh but to update the size property. I wonder if we could update it in `onDataProviderPageLoaded` instead.

Reply:
> I wonder if the better way would be to update it in onDataProviderPageLoaded instead.

Maybe in an overridden `onDataProviderPageLoaded` in array-data-provider-mixin (we probably don't want to have `size` auto-assigned when a `dataProvider` is used).

To me the PR is ok as it is. It does increase the array data-provider calls from 1 to 2 on multi-sort but that's probably not an issue and as you said, it works that way with single sorting.